File: 2011_03_16/caixa.py
class ValorNaoPermitido(Exception):
    pass

class Caixa(object):
    notas = [100, 50, 20, 10]
    
    def valor_e_permitido(self, valor):
        return valor % 10 != 0 or valor < 1

    def sacar(self, valor):
        if self.valor_e_permitido(valor):
            raise ValorNaoPermitido
            
        saque = []
        
        for nota in self.notas:
            while valor >= nota:
                valor = valor - nota
                saque.append(nota)
                
        return saque
           
    def como_dicionario(self, notas_sacadas):
        notas_dict = {} 
        
        for nota in self.notas:
            if nota in notas_sacadas:
                notas_dict[nota] = notas_sacadas.count(nota)
               
        return notas_dict

File: 2015/setembro/2015_09_01/caixa.go
package caixa

import (
	"errors"
)

var notasDisponiveis = [4]int{100, 50, 20, 10}

func Saque (valor int) ([]int, error) {
	notasSacadas := []int{}
	valorRestante := valor

	for i := 0; i < len(notasDisponiveis); i++ {
		for valorRestante > 0 && valorRestante >= notasDisponiveis[i] {
			if notasDisponiveis[i] <= valorRestante && valorRestante > 0 {
				notasSacadas = append(notasSacadas, notasDisponiveis[i])
				valorRestante -= notasDisponiveis[i]
			} else {
				break
			}
		}
	}

	if valorRestante > 0 {
		return []int{}, errors.New("valor invalido")
	} else {
		return notasSacadas, nil
	}
}